R/GeospatialFunctions.R

```diff
@@ -0,0 +1,607 @@
+  # together.
+  
+  repeat {
+    
+    query <- urltools::param_set(baseurl, key = "geometry", value = bbox) %>%
+      urltools::param_set(key = "inSR", value = epsg) %>%
+      # Total of 2000 features at a time...
+      urltools::param_set(key = "resultRecordCount", value = 500) %>%
+      # ... starting at the "offset":
+      urltools::param_set(key = "resultOffset", value = offset) %>%
+      urltools::param_set(key = "spatialRel", value = "esriSpatialRelIntersects") %>%
+      urltools::param_set(key = "f", value = "geojson") %>%
+      urltools::param_set(key = "outFields", value = "*") %>%
+      urltools::param_set(key = "geometryType", value = "esriGeometryEnvelope") %>%
+      urltools::param_set(key = "returnGeometry", value = "true") %>%
+      urltools::param_set(key = "returnTrueCurves", value = "false") %>%
+      urltools::param_set(key = "returnIdsOnly", value = "false") %>%
+      urltools::param_set(key = "returnCountOnly", value = "false") %>%
+      urltools::param_set(key = "returnZ", value = "false") %>%
+      urltools::param_set(key = "returnM", value = "false") %>%
+      urltools::param_set(key = "returnDistinctValues", value = "false") %>%
+      urltools::param_set(key = "returnExtentOnly", value = "false") %>%
+      urltools::param_set(key = "featureEncoding", value = "esriDefault")
+    
+    # Fetch features within the offset window and append to list:
+    features <- suppressMessages(suppressWarnings({tryCatch({
+      geojsonsf::geojson_sf(query)
+    }, error = function(e) {
+      NULL
+    })}))
+    
+    
+    # Exit loop if no more features or error occurred
+    if (is.null(features) || nrow(features) == 0) {
+      break
+    }
+    
+    all_features <- c(all_features, list(features))
+    # once done, change offset by 2000 features:
+    offset <- offset + 500
+    
+    if(offset == 4000){print("Your TADA data covers a large spatial range. The ATTAINS pull may take a while.")}
+    
+  }
+  
+  all_features <- dplyr::bind_rows(all_features)
+  
+  return(all_features)
+  
+}
+
+
+#' TADA_GetATTAINS
+#' 
+#' Link catchment-based ATTAINS assessment unit data to Water Quality Portal observations, often imported via `TADA_DataRetrieval()`. This function returns the same raw objects that are mapped in `TADA_ViewATTAINS()`.
+#' 
+#' @param data A dataframe created by `TADA_DataRetrieval()` or the sf equivalent made by `TADA_MakeSpatial()`.
+#' @param return Whether to return the associated ATTAINS_catchments, ATTAINS_lines, ATTAINS_points, and ATTAINS_polygons shapefile objects. TRUE (yes, return) or FALSE (no, do not return). All ATTAINS features are in WGS84 (crs = 4326).
+#' 
+#' @return A modified `TADA_DataRetrieval()` dataframe with additional columns associated with the ATTAINS assessment unit data. Or, if return = TRUE, a list containing that same data frame plus the raw ATTAINS features associated with those observations.
```

Comment:
I'm confused about the return argument here - does the inclusion of 'additional ATTAINS features' make it into a sf object? does setting FALSE decrease run time? I think the foundation of my questions is "why" wouldn't someone want all of the ATTAINS info?

Later comment: if set to FALSE, this returns a dataframe, but if set to TRUE, this returns a list of dataframes?

Reply:
Also, since `return` is an R functionality that is used in function dev, this makes reading this function pretty difficult because there are also early returns in this fx. I'd suggest re-naming this argument to avoid confusion.